src/H5TSint.c

```diff
@@ -0,0 +1,577 @@
+#include "H5TSmodule.h" /* This source code file is part of the H5TS module */
+
+/***********/
+/* Headers */
+/***********/
+#include "H5private.h"   /* Generic Functions                   */
+#include "H5CXprivate.h" /* API Contexts                        */
+#include "H5Epkg.h"      /* Error handling                      */
+#include "H5TSpkg.h"     /* Threadsafety                        */
+
+#ifdef H5_HAVE_THREADSAFE
+
+/****************/
+/* Local Macros */
+/****************/
+
+#ifdef H5_HAVE_C11_THREADS
+#define H5TS_ONCE_INIT_FUNC H5TS__c11_first_thread_init
+#else
+#ifdef H5_HAVE_WIN_THREADS
+#define H5TS_ONCE_INIT_FUNC H5TS__win32_process_enter
+#else
+#define H5TS_ONCE_INIT_FUNC H5TS__pthread_first_thread_init
+#endif /* H5_HAVE_WIN_THREADS */
+#endif
+
+/******************/
+/* Local Typedefs */
+/******************/
+
+/* Per-thread info */
+typedef struct H5TS_thread_info_t {
+    uint64_t            id;               /* Unique ID for each thread */
+    struct H5CX_node_t *api_ctx_node_ptr; /* Pointer to an API context node */
+    H5E_stack_t         err_stack;        /* Error stack */
+} H5TS_thread_info_t;
+
+/* An H5TS_tinfo_node_t is a thread info that is available for reuse */
+typedef struct H5TS_tinfo_node_t {
+    struct H5TS_tinfo_node_t *next;
+    H5TS_thread_info_t        info;
+} H5TS_tinfo_node_t;
+
+/********************/
+/* Local Prototypes */
+/********************/
+static H5TS_tinfo_node_t *H5TS__tinfo_create(void);
+
+/*********************/
+/* Package Variables */
+/*********************/
+
+/* Per-thread info */
+H5TS_key_t H5TS_thrd_info_key_g;
+
+/*****************************/
+/* Library Private Variables */
+/*****************************/
+
+/*******************/
+/* Local Variables */
+/*******************/
+
+/* Has threadsafety code been initialized? */
+static H5TS_once_t H5TS_first_init_s = H5TS_ONCE_INITIALIZER;
+
+/* Pointer to first free thread info record or NULL. */
+static H5TS_tinfo_node_t *H5TS_tinfo_next_free_s = NULL;
+static uint64_t           H5TS_next_thrd_id_s    = 0;
+
+/* Mutex for access to H5TS_tinfo_next_free_s and H5TS_next_thrd_id_s */
+static H5TS_mutex_t H5TS_tinfo_mtx_s;
+
+/*-------------------------------------------------------------------------
+ * Function: H5TS__init
+ *
+ * Purpose:  Initialize the H5TS interface
+ *
+ * Return:   Non-negative on success / Negative on failure
+ *
+ *-------------------------------------------------------------------------
+ */
+herr_t
+H5TS__init(void)
+{
+    herr_t ret_value = SUCCEED; /* Return value */
+
+    FUNC_ENTER_PACKAGE_NAMECHECK_ONLY
+
+    /* Initialize the global API lock info */
+    if (H5_UNLIKELY(H5TS_mutex_init(&H5TS_api_info_p.api_mutex, H5TS_MUTEX_TYPE_RECURSIVE) < 0))
+        HGOTO_DONE(FAIL);
+    H5TS_api_info_p.lock_count = 0;
+    H5TS_atomic_init_uint(&H5TS_api_info_p.attempt_lock_count, 0);
+
+    /* Initialize per-thread library info */
+    if (H5_UNLIKELY(H5TS__tinfo_init() < 0))
```

Comment:
H5TS__init() is only called by one thread right? How do the other threads call H5TS__tinfo_init()?